pkg/gits/helpers.go

```diff
@@ -384,192 +379,185 @@ func PushRepoAndCreatePullRequest(dir string, gitInfo *GitRepository, base strin
-			if err != nil {
-				return "", "", nil, errors.Wrap(err, "failed to create temp dir")
-			}
-
-			err = os.MkdirAll(dir, util.DefaultWritePermissions)
-			if err != nil {
-				return "", "", nil, fmt.Errorf("Failed to create directory %s due to %s", dir, err)
-			}
-			cloneGitURL, err := gitter.CreatePushURL(repo.CloneURL, &userDetails)
-			if err != nil {
-				return "", "", nil, errors.Wrapf(err, "failed to get clone URL from %s and user %s", repo.CloneURL, username)
-			}
-			err = gitter.Clone(cloneGitURL, dir)
-			if err != nil {
-				return "", "", nil, errors.WithStack(err)
-			}
-			err = gitter.FetchBranch(dir, "origin")
-			if err != nil {
-				return "", "", nil, errors.Wrapf(err, "fetching from %s", cloneGitURL)
-			}
-			err = gitter.SetRemoteURL(dir, "upstream", gitURL)
-			if err != nil {
-				return "", "", nil, errors.Wrapf(err, "setting remote upstream %q in forked environment repo", gitURL)
-			}
-			if configGitFn != nil {
-				err = configGitFn(dir, gitInfo, gitter)
-				if err != nil {
-					return "", "", nil, errors.WithStack(err)
-				}
-			}
-			branchName, err := computeBranchName(baseRef, branchName, dir, gitter)
-			if err != nil {
-				return "", "", nil, errors.WithStack(err)
-			}
-			if branchName != "master" {
-				err = gitter.CreateBranch(dir, branchName)
-				if err != nil {
-					return "", "", nil, errors.WithStack(err)
-				}
-
-				err = gitter.Checkout(dir, branchName)
-				if err != nil {
-					return "", "", nil, errors.WithStack(err)
-				}
-			}
-			err = gitter.ResetToUpstream(dir, baseRef)
-			if err != nil {
-				return "", "", nil, errors.Wrapf(err, "resetting forked branch %s to upstream version", baseRef)
-			}
-			return dir, baseRef, gitInfo, nil
+			log.Logger().Infof("Forked Git repository to %s\n", util.ColorInfo(gitInfo.HTMLURL))
 		}
+	} else {
+		// the normal workflow for a non-forked model is to use "origin" not  upstream so let's emulate that here
+		gitInfo = upstreamInfo
 	}
 
-	// now lets clone the fork and pull it...
-	exists, err := util.FileExists(dir)
+	cloneGitURL, err := gitter.CreatePushURL(gitInfo.CloneURL, &userDetails)
 	if err != nil {
-		return "", "", nil, errors.Wrapf(err, "failed to check if directory %s exists", dir)
+		return "", "", nil, errors.Wrapf(err, "failed to get clone URL from %s and user %s", gitURL, username)
 	}
 
-	if exists {
-		if configGitFn != nil {
-			err = configGitFn(dir, gitInfo, gitter)
-			if err != nil {
-				return "", "", nil, errors.WithStack(err)
-			}
-		}
-		// lets check the gitter remote URL is setup correctly
-		err = gitter.SetRemoteURL(dir, "origin", gitURL)
+	// Prepare the git repo
+
+	if !dirExists {
+		// If the directory doesn't already exist, create it
+		err := os.MkdirAll(dir, util.DefaultWritePermissions)
 		if err != nil {
-			return "", "", nil, errors.WithStack(err)
+			return "", "", nil, fmt.Errorf("failed to create directory %s due to %s", dir, err)
 		}
-		err = gitter.Stash(dir)
+	}
+
+	stashed := false
```

Comment:
So when are there really stashed changes in this checkout? I would not expect this really from a "fork and pull" function? Does this function not really try to do multiple things?

Reply:
We can rename the function I guess. What do you suggest?
 
The functionality is required to pull into an existing dir and not clobber existing files.

Reply:
This is covered in the comments.

Reply:
> The functionality is required to pull into an existing dir and not clobber existing files.

Sure, I guess I would have done this differently. I'd check whether the dir does not exist or is not a git repo and in this case do a standard fork and clone and if it exist I'd add the remote and do the other bits required. 

Maybe a different name could be `ForkAndPullIntoLocalRepo`!?